pkg/controllers/workgenerator/controller.go

```diff
@@ -89,29 +95,50 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		return ctrl.Result{}, err
 	}
 
-	// generate and apply the updated works
-	syncErr := r.syncAllWork(ctx, resourceBinding.DeepCopy())
+	// make a copy of the binding so that it won't pollute the cache
+	binding := resourceBinding.DeepCopy()
+	workUpdated := false
+	// list all the corresponding works
+	works, syncErr := r.listAllWorksAssociated(ctx, binding)
+	if syncErr == nil {
+		// generate and apply the workUpdated works if we have all the works
+		workUpdated, syncErr = r.syncAllWork(ctx, binding, works)
+	}
+
 	if syncErr != nil {
 		klog.ErrorS(syncErr, "Failed to sync all the works", "resourceBinding", bindingRef)
-		resourceBinding.SetConditions(metav1.Condition{
+		binding.SetConditions(metav1.Condition{
```

Comment:
CRP status implementation is if it hits any api server error, it won't set the status, keep the old status there and then do the retry.

It will be simpler compared with this way.